pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -0,0 +1,346 @@
+
+    # Next, extract B. Can do from any of the C, just need to be careful in
+    # case one of the elements of A is 0.
+    if not qml.math.isclose(A[0, 0], 0.0, atol=1e-8):
+        B = C1 / A[0, 0]
+    else:
+        B = C2 / A[0, 1]
+
+    return A, B
+
+
+def _extract_su2su2_prefactors(U, V):
+    """U, V are SU(4) matrices for which there exists A, B, C, D such that
+    (A \otimes B) V (C \otimes D) = U. The problem is to find A, B, C, D in SU(2)
+    in an analytic and fully differentiable manner.
+
+    This decomposition is possible when U and V are in the same double coset of
+    SU(4), meaning there exists G, H in SO(4) s.t. G (Edag V E) H = (Edag U
+    E). This is guaranteed here by how V was constructed using the
+    _select_rotation_angles method. Then, we can use the fact that E SO(4) Edag
+    gives us something in SU(2) x SU(2) to give A, B, C, D.
+    """
+
+    # A lot of the work here happens in the magic basis. Essentially, we
+    # don't look explicitly at some U = G V H, but rather at
+    #     E^\dagger U E = G E^\dagger V E H
+    # so that we can recover
+    #     U = (E G E^\dagger) V (E H E^\dagger) = (A \otimes B) V (C \otimes D).
+    # There is some math in the paper explaining how when we define U in this way,
+    # we can simultaneously diagonalize functions of U and V to ensure they are
+    # in the same coset and recover the decomposition.
+
+    u = qml.math.linalg.multi_dot([Edag, U, E])
+    v = qml.math.linalg.multi_dot([Edag, V, E])
+
+    uuT = qml.math.dot(u, qml.math.T(u))
+    vvT = qml.math.dot(v, qml.math.T(v))
+
+    # First, we find a matrix p (hopefully) in SO(4) s.t. p^T u u^T p is diagonal.
+    # Since uuT is complex and symmetric, both its real / imag parts share a set
+    # of real-valued eigenvectors.
+    ev_p, p = qml.math.linalg.eig(uuT.real)
+
+    # We also do this for v, i.e., find q (hopefully) in SO(4) s.t. q^T v v^T q is diagonal.
+    ev_q, q = qml.math.linalg.eig(vvT.real)
+
+    # If determinant of p is not 1, it is in O(4) but not SO(4), and has
+    # determinant -1. We can transform it to SO(4) by simply negating one
+    # of the columns.
+    if not qml.math.isclose(qml.math.linalg.det(p), 1.0):
+        p[:, -1] = -p[:, -1]
+
+    # Next, we are going to reorder the columns of q so that the order of the
+    # eigenvalues matches those of p.
+    p_product = qml.math.linalg.multi_dot([qml.math.T(p), uuT, p])
+    q_product = qml.math.linalg.multi_dot([qml.math.T(q), vvT, q])
+
+    p_diag = qml.math.diag(p_product)
+    q_diag = qml.math.diag(q_product)
+
+    new_q_order = []
+
+    for idx, eigval in enumerate(p_diag):
+        are_close = [qml.math.isclose(x, eigval) for x in q_diag]
+
+        if any(are_close):
+            new_q_order.append(qml.math.argmax(are_close))
+
+    # Get the permutation matrix needed to reshuffle the columns
+    q_perm = _perm_matrix_from_sequence(new_q_order)
+    q = qml.math.linalg.multi_dot([q, qml.math.T(q_perm)])
```

Comment:
Wait, could this part be avoided by simply using indexing to do the re-ordering?
```python
q = q[:, new_q_order]
```